src/main/java/io/supertokens/inmemorydb/queries/TOTPQueries.java

```diff
@@ -0,0 +1,179 @@
+            pst.setString(2, device.userId);
+            pst.setString(3, device.secretKey);
+            pst.setInt(4, device.period);
+            pst.setInt(5, device.skew);
+            pst.setBoolean(6, device.verified);
+        });
+    }
+
+    public static int markDeviceAsVerified(Start start, String userId, String deviceName)
+            throws StorageQueryException, SQLException {
+        String QUERY = "UPDATE " + Config.getConfig(start).getTotpUserDevicesTable()
+                + " SET verified = true WHERE user_id = ? AND device_name = ?;";
+        return update(start, QUERY, pst -> {
+            pst.setString(1, userId);
+            pst.setString(2, deviceName);
+        });
+    }
+
+    public static int deleteDevice(Start start, String userId, String deviceName)
+            throws StorageQueryException, SQLException {
+        String QUERY = "DELETE FROM " + Config.getConfig(start).getTotpUserDevicesTable()
+                + " WHERE user_id = ? AND device_name = ?;";
+
+        return update(start, QUERY, pst -> {
+            pst.setString(1, userId);
+            pst.setString(2, deviceName);
+        });
+    }
+
+    public static int updateDeviceName(Start start, String userId, String oldDeviceName, String newDeviceName)
+            throws StorageQueryException, SQLException {
+        String QUERY = "UPDATE " + Config.getConfig(start).getTotpUserDevicesTable()
+                + " SET device_name = ? WHERE user_id = ? AND device_name = ?;";
+
+        return update(start, QUERY, pst -> {
+            pst.setString(1, newDeviceName);
+            pst.setString(2, userId);
+            pst.setString(3, oldDeviceName);
+        });
+    }
+
+    public static TOTPDevice[] getDevices(Start start, String userId)
+            throws StorageQueryException, SQLException {
+        String QUERY = "SELECT * FROM " + Config.getConfig(start).getTotpUserDevicesTable()
+                + " WHERE user_id = ?;";
+
+        return execute(start, QUERY, pst -> pst.setString(1, userId), result -> {
+            List<TOTPDevice> devices = new ArrayList<>();
+            while (result.next()) {
+                devices.add(TOTPDeviceRowMapper.getInstance().map(result));
+            }
+
+            return devices.toArray(TOTPDevice[]::new);
+        });
+    }
+
+    public static int insertUsedCode(Start start, TOTPUsedCode code) throws SQLException, StorageQueryException {
+        String QUERY = "INSERT INTO " + Config.getConfig(start).getTotpUsedCodesTable()
+                + " (user_id, code, is_valid_code, expiry_time_ms) VALUES (?, ?, ?, ?);";
+
+        return update(start, QUERY, pst -> {
+            pst.setString(1, code.userId);
+            pst.setString(2, code.code);
+            pst.setBoolean(3, code.isValidCode);
+            pst.setLong(4, code.expiryTime);
+        });
+    }
+
+    public static TOTPUsedCode[] getUsedCodes(Start start, String userId) throws SQLException, StorageQueryException {
```

Comment:
this function needs to only return non expired codes

Reply:
Also add a test for this in the storage layer part

Reply:
done. See `getNonExpiredUsedCodesTest`